api/envoy/service/ratelimit/v3/rls.proto

```diff
@@ -49,6 +49,8 @@ message RateLimitRequest {
 
   // Rate limit requests can optionally specify the number of hits a request adds to the matched
   // limit. If the value is not set in the message, a request increases the matched limit by 1.
+  // This value can be modified by setting "envoy.ratelimit:hits_addend" in the dynamic metadata
+  // for a given request.
```

Comment:
In most of case, we prefer do similar things in the `FilterState` rather than dynamic metadata? Is there any reason prefer the dynamic metadata?

cc @mattklein123

Reply:
I based this off of `envoy_lb` style decisions which can be made via `DynamicMetadata`. I don't have super strong feelings either way, but there are more easier ways to set `DynamicMetadata` via other filters so I slightly prefer this method.

Reply:
`envoy.lb` use the metadata because the subset lb construct the match structure from the endpoint metadata.

Then, in this scenario, I think filter state should be preferred for these typed options.

Reply:
in this case we use dynamic metadata as it allows to feed data in without the source filter being aware that it is for rate limit. 
For example, we can use ext-auth server to add the metadata for the rate limit filter.

Reply:
> in this case we use dynamic metadata as it allows to feed data in without the source filter being aware that it is for rate limit. For example, we can use ext-auth server to add the metadata for the rate limit filter.

How? As far as I know, the ext_authz could inject the dynamic metadata unser namespace `envoy.filters.http.ext_authz`?

Reply:
I don't know that I have a super strong opinion on this, but I probably should have asked previously: How exactly is this going to be used? I don't quite understand who is going to be setting the metadata here? If @wbpcode thinks filter state is better that is fine with me.

/wait

Reply:
The use-case is that we want to dynamically increment the `hits_addend` value based on the contents of the body. We have a filter to process the body and then write the value to the `dynamic_metadata`. Maybe I'm not understanding the distinction between `dynamic_metadata` and `filter_state`. Typically I've used `dynamic_metadata` to internally pass data between multiple filters, but it seems like there's nuance here I'm missing.

Reply:
Here are some discussion about the filter state and dynamic metadata https://github.com/envoyproxy/envoy/issues/25130

Basically, we prefer to use filter state for non-proto/JSON-like data.

Reply:
Alright, I plan to use the [`UInt32Accessor`](https://github.com/envoyproxy/envoy/blob/693a94620a844f8ffde5897771bee4888dc013b1/envoy/stream_info/uint32_accessor.h) class. Does that make sense?

Reply:
@wbpcode does that sound correct?

Reply:
> We have a filter to process the body and then write the value to the dynamic_metadata

Would this also allow forwarding the call to the downstream service and e.g. taking something off the response header for `hits_addend` before hitting the rate limiter, as described here? https://github.com/envoyproxy/ratelimit/issues/636#issuecomment-2201132971